### PR TITLE
feat(shwap):Add eds streaming

### PR DESCRIPTION
**Pull request by walldiss**
Add streaming: read/write for files. Read is implemented only for Rsmt2d file by `ReadFrom` method. There is no usecase so farm to directly read for ods file, so `ReadFrom` is not added to eds.Accessor interface

### CODE REVIEW

File: share/new_eds/rsmt2d.go [hunk 1]
```diff
@@ -11,20 +12,20 @@ import (
 	"github.com/celestiaorg/celestia-node/share/shwap"
 )
 
-var _ Accessor = Rsmt2D{}
+// var _ Accessor = Rsmt2D{}
```
Comment (cristaloleg) [inline share/new_eds/rsmt2d.go:15]: Remove?

File: store/file/square.go [hunk 2]
```diff
@@ -141,3 +150,34 @@ func oppositeAxis(axis rsmt2d.Axis) rsmt2d.Axis {
...
+}
+
+type odsWriter struct {
+	ods square
+	// current is the amount of Shares stored in square that have been written by odsWriter. When
+	// current reaches total, odsWriter will prevent further reads by returning io.EOF
+	current, total int
+}
+
+func (w *odsWriter) WriteTo(writer io.Writer, limit int) (int, error) {
```
Comment (cristaloleg) [inline store/file/square.go:168]: `WriteTo` is a method from a well-known interface https://pkg.go.dev/io#WriterTo

Quite often breaking such unspoken rules might result in a future incompatibilities between implementation and stdlib interfaces. Just saying.

File: share/new_eds/reader.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+func NewSharesReader(odsSize int, getShare func(rowIdx, colIdx int) ([]byte, error)) *BufferedReader {
+	return &BufferedReader{
+		getShare: getShare,
+		buf:      bytes.NewBuffer(nil),
+		odsSize:  odsSize,
+		total:    odsSize * odsSize,
+	}
+}
+
+// BufferedReader will read Shares from inMemOds into the buffer.
```
Comment (Wondertan) [inline share/new_eds/reader.go:21]:
```suggestion
// BufferedReader will read Shares from getShare function into the buffer.
```

File: store/file/ods_test.go [hunk 4]
```diff
@@ -62,7 +62,9 @@ func TestODSFile(t *testing.T) {
 	t.Cleanup(cancel)
 
 	ODSSize := 8
```
Comment (Wondertan) [inline store/file/ods_test.go:64]: It is worth increasing the size so it's bigger than the buffer. Ideally, we would do specify MaxODS to the test suite so it tests everything for all the sizes until the max given. It will take more time, yes, but we have high assurance that all sizes work correctly.

File: store/file/ods.go [hunk 5]
```diff
@@ -288,3 +306,24 @@ func (f *ODSFile) axis(ctx context.Context, axisType rsmt2d.Axis, axisIdx int) (
...
+		total:   int64(odsSize*odsSize*shareSize + offset),
+	}
+}
+
+func (w *fileReader) Read(p []byte) (int, error) {
+	if w.current >= w.total {
+		return 0, io.EOF
+	}
+
+	return w.r.ReadAt(p, w.current)
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply store/file/ods.go:328]: Well, nvm tests aren't working

File: share/new_eds/reader.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	if !errors.Is(err, io.ErrUnexpectedEOF) && !errors.Is(err, io.EOF) {
+		return n, fmt.Errorf("unexpected error reading from buf: %w", err)
+	}
+
+	written := n
+	for r.current < r.total {
+		rowIdx, colIdx := r.current/r.odsSize, r.current%r.odsSize
+		share, err := r.getShare(rowIdx, colIdx)
+		if err != nil {
+			return 0, fmt.Errorf("get share; %w", err)
```
Comment (Wondertan) [inline share/new_eds/reader.go:52]:
```suggestion
			return 0, fmt.Errorf("get share: %w", err)
```

File: share/new_eds/close_once.go [hunk 7]
```diff
@@ -11,16 +12,16 @@ import (
...
 var errAccessorClosed = errors.New("accessor is closed")
 
 type closeOnce struct {
-	f      AccessorCloser
+	f      AccessorStreamer
 	closed atomic.Bool
 }
 
-func WithClosedOnce(f AccessorCloser) AccessorCloser {
+func WithClosedOnce(f AccessorStreamer) AccessorStreamer {
```
[the remark this answers is not in the document]
Comment (walldiss) [reply share/new_eds/close_once.go:24]: Will be used later. It is meant for storage

File: share/new_eds/rsmt2d.go [hunk 8]
```diff
@@ -89,10 +90,36 @@ func (eds Rsmt2D) RowNamespaceData(
...
+		return eds.GetCell(uint(rowIdx), uint(colIdx)), nil
+	}
+	odsSize := int(eds.Width() / 2)
+	reader := NewSharesReader(odsSize, getShare)
+	return reader, nil
+}
+
+// Rsmt2DFromShares reads shares from the provided reader and constructs an Extended Data Square.
+// Provided reader should contain shares in row-major order.
+func Rsmt2DFromShares(shares []share.Share, odsSize int) (Rsmt2D, error) {
```
[the remark this answers is not in the document]
Comment (walldiss) [reply share/new_eds/rsmt2d.go:113]: It doesn't need reader. It is jsut a constructor from share.

Comment updated

File: share/new_eds/accessor.go [hunk 9]
```diff
@@ -27,17 +27,24 @@ type Accessor interface {
...
+// AccessorStreamer is an interface that groups Accessor and Streamer interfaces.
+type AccessorStreamer interface {
 	Accessor
+	Streamer
+}
+
+type Streamer interface {
+	// Reader returns binary reader for the file (ODS) shares. It should read the shares from the
+	// ODS part of the square row by row.
+	Reader() (io.Reader, error)
```
[the remark this answers is not in the document]
Comment (walldiss) [reply share/new_eds/accessor.go:39]: Sync: decided to keep Reader as part of Streamer interface and use it only for file implementation for now.

[the remark this answers is not in the document]
Comment (walldiss) [reply share/new_eds/accessor.go:39]: Sync: ctx not needed if we don't include it into Accessor interface

File: store/file/ods.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (walldiss) [reply store/file/ods.go:328]: removed that code

File: share/new_eds/accessor.go [hunk 10]
```diff
@@ -27,17 +27,24 @@ type Accessor interface {
...
-type accessorCloser struct {
+type accessorStreamer struct {
 	Accessor
-	io.Closer
+	Streamer
 }
 
-func WithCloser(a Accessor, c io.Closer) AccessorCloser {
-	return &accessorCloser{a, c}
+func AccessorAndStreamer(a Accessor, s Streamer) AccessorStreamer {
```
Comment (Wondertan) [inline share/new_eds/accessor.go:48]: When we discussed the idiomatic approach, I think you wanted to find `iotest` pkg, which has things like [`TimeoutReader`](https://pkg.go.dev/testing/iotest#TimeoutReader)

File: share/new_eds/proofs_cache.go [hunk 11]
```diff
@@ -211,21 +212,28 @@ func (c *proofsCache) Shares(ctx context.Context) ([]share.Share, error) {
...
-		if err != nil {
-			return nil, err
-		}
-		ax.half = half
+	half, err := c.AxisHalf(ctx, axisType, axisIdx)
+	if err != nil {
+		return nil, err
 	}
 
-	shares, err := ax.half.Extended()
```
Comment (Wondertan) [inline share/new_eds/proofs_cache.go:228]: What's the reason for this change? I don't follow.

**Review by Wondertan — APPROVED** [review]

File: share/new_eds/accessor.go (hunk 10 above)
Comment (walldiss) [reply share/new_eds/accessor.go:48]: // TimeoutReader returns [ErrTimeout](https://pkg.go.dev/testing/iotest#ErrTimeout) on the second read with no data. Subsequent calls to read succeed.

Not sure how it is relevant

File: share/new_eds/proofs_cache.go (hunk 11 above)
Comment (walldiss) [reply share/new_eds/proofs_cache.go:228]: It caches half axis independently inside AxisHalf call
